File: ruby/lib/data_structures/min_heap.rb
# frozen_string_literal: true

module DSA
  module DataStructures
    # Binary min-heap implementation.
    # Time: O(log n) insert/extract, O(1) peek
    # Space: O(n)
    class MinHeap
      def initialize(&comparator)
        @heap = []
        @comparator = comparator || ->(a, b) { a <=> b }
      end

      def insert(value)
        @heap << value
        sift_up(@heap.size - 1)
        self
      end

      def extract_min
        return nil if @heap.empty?

        min = @heap[0]
        last = @heap.pop

        unless @heap.empty?
          @heap[0] = last
          sift_down(0)
        end

        min
      end

      def peek
        @heap.first
      end

      def empty?
        @heap.empty?
      end

      def size
        @heap.size
      end

      def clear
        @heap.clear
        self
      end

      def to_a
        @heap.dup
      end

      private

      def sift_up(index)
        while index > 0
          parent_index = (index - 1) / 2
          break if @comparator.call(@heap[index], @heap[parent_index]) >= 0

          swap(index, parent_index)
          index = parent_index
        end
      end

      def sift_down(index)
        loop do
          left_child = 2 * index + 1
          right_child = 2 * index + 2
          smallest = index

          if left_child < @heap.size && @comparator.call(@heap[left_child], @heap[smallest]) < 0
            smallest = left_child
          end

          if right_child < @heap.size && @comparator.call(@heap[right_child], @heap[smallest]) < 0
            smallest = right_child
          end

          break if smallest == index

          swap(index, smallest)
          index = smallest
        end
      end

      def swap(i, j)
        @heap[i], @heap[j] = @heap[j], @heap[i]
      end
    end
  end
end

File: typescript/src/data-structures/min-heap.ts
/**
 * Binary min-heap implementation.
 * Time: O(log n) insert/extract, O(1) peek
 * Space: O(n)
 */

type Comparator<T> = (a: T, b: T) => number;

export class MinHeap<T> {
  private heap: T[] = [];
  private compare: Comparator<T>;

  constructor(comparator?: Comparator<T>) {
    this.compare = comparator || ((a, b) => {
      if (a < b) return -1;
      if (a > b) return 1;
      return 0;
    });
  }

  insert(value: T): void {
    this.heap.push(value);
    this.siftUp(this.heap.length - 1);
  }

  extractMin(): T | undefined {
    if (this.heap.length === 0) {
      return undefined;
    }

    const min = this.heap[0];
    const last = this.heap.pop()!;

    if (this.heap.length > 0) {
      this.heap[0] = last;
      this.siftDown(0);
    }

    return min;
  }

  peek(): T | undefined {
    return this.heap[0];
  }

  isEmpty(): boolean {
    return this.heap.length === 0;
  }

  size(): number {
    return this.heap.length;
  }

  clear(): void {
    this.heap = [];
  }

  toArray(): T[] {
    return [...this.heap];
  }

  private siftUp(index: number): void {
    while (index > 0) {
      const parentIndex = Math.floor((index - 1) / 2);
      if (this.compare(this.heap[index], this.heap[parentIndex]) >= 0) {
        break;
      }
      this.swap(index, parentIndex);
      index = parentIndex;
    }
  }

  private siftDown(index: number): void {
    const length = this.heap.length;

    while (true) {
      const leftChild = 2 * index + 1;
      const rightChild = 2 * index + 2;
      let smallest = index;

      if (leftChild < length && this.compare(this.heap[leftChild], this.heap[smallest]) < 0) {
        smallest = leftChild;
      }

      if (rightChild < length && this.compare(this.heap[rightChild], this.heap[smallest]) < 0) {
        smallest = rightChild;
      }

      if (smallest === index) {
        break;
      }

      this.swap(index, smallest);
      index = smallest;
    }
  }

  private swap(i: number, j: number): void {
    [this.heap[i], this.heap[j]] = [this.heap[j], this.heap[i]];
  }
}
